common/kvstore/leveldb/metrics.go

```diff
@@ -0,0 +1,475 @@
+	ticker := time.NewTicker(mc.config.CollectionInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-mc.stopChan:
+			return
+		case <-ticker.C:
+			mc.collectMetrics()
+		}
+	}
+}
+
+func (mc *MetricsCollector) collectMetrics() {
+	var stats leveldb.DBStats
+	if err := mc.db.Stats(&stats); err != nil {
+		mc.logger.Error("Failed to collect database stats", "error", err)
+		return
+	}
+
+	mc.mu.Lock()
+	defer mc.mu.Unlock()
+
+	// Calculate time-based deltas
+	timeDelta := time.Since(mc.lastCollection).Seconds()
+	if timeDelta == 0 {
+		return // Avoid division by zero
+	}
+
+	// Process compaction metrics
+	mc.processMetrics(&stats, timeDelta)
+
+	// Check for performance degradation
+	mc.checkDegradation(&stats)
+
+	// Update state
+	mc.lastStats = stats
+	mc.lastCollection = time.Now()
+}
+
+func (mc *MetricsCollector) processMetrics(stats *leveldb.DBStats, timeDelta float64) {
+	// Calculate compaction latencies
+	if !mc.lastCollection.IsZero() && len(mc.lastStats.LevelDurations) > 0 {
+		for level, duration := range stats.LevelDurations {
+			if level < len(mc.lastStats.LevelDurations) {
+				prevDuration := mc.lastStats.LevelDurations[level]
+				deltaDuration := duration - prevDuration
+				if deltaDuration > 0 {
+					compactionLatency.WithLabelValues(getLevelName(level), mc.config.Name).Observe(deltaDuration.Seconds())
+				}
+			}
+		}
+	}
+
+	// Calculate total compaction time (cumulative)
+	var totalDuration time.Duration
+	for _, duration := range stats.LevelDurations {
+		totalDuration += duration
+	}
+
+	totalCompactionTime.WithLabelValues(mc.config.Name).Set(totalDuration.Seconds())
+
+	// Calculate throughput metrics
+	if prevStats := mc.lastStats; prevStats.LevelRead != nil {
+		readDelta := stats.LevelRead.Sum() - prevStats.LevelRead.Sum()
+		writeDelta := stats.LevelWrite.Sum() - prevStats.LevelWrite.Sum()
+
+		compactionThroughput.WithLabelValues("read", mc.config.Name).Set(float64(readDelta) / timeDelta)
+		compactionThroughput.WithLabelValues("write", mc.config.Name).Set(float64(writeDelta) / timeDelta)
+	}
+
+	// Update compaction counters for each level
+	compactionCount.WithLabelValues("memory", mc.config.Name).Set(float64(stats.MemComp))
+	compactionCount.WithLabelValues("level0", mc.config.Name).Set(float64(stats.Level0Comp))
+	compactionCount.WithLabelValues("nonlevel0", mc.config.Name).Set(float64(stats.NonLevel0Comp))
+	compactionCount.WithLabelValues("seek", mc.config.Name).Set(float64(stats.SeekComp))
+
+	// Process resource metrics
+	openTableCount.WithLabelValues(mc.config.Name).Set(float64(stats.OpenedTablesCount))
+	blockCacheSize.WithLabelValues(mc.config.Name).Set(float64(stats.BlockCacheSize))
+
+	// Process performance metrics
+	if !mc.lastCollection.IsZero() {
+		readDelta := float64(stats.IORead - mc.lastStats.IORead)
+		writeDelta := float64(stats.IOWrite - mc.lastStats.IOWrite)
+
+		readThroughput.WithLabelValues(mc.config.Name).Set(readDelta / timeDelta)
```

Comment:
Do these two throughput include the compaction throughput (that's computed based on stats.LevelWrite.Sum() above)?

Reply:
Yeah, IORead and IOWrite is measuring the total read/write at the disk layer.